In [1]:
import pandas as pd
import numpy as np
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)
Turkey = pd.read_excel("ASHTURR5.xlsx")
Austria = pd.read_excel("ASHAUTR5.xlsx")
Germany = pd.read_excel("ASHDEUR5.xlsx")
Egypt = pd.read_excel("ASHEGYR5.xlsx")
France = pd.read_excel("ASHFRAR5.xlsx")
Iran = pd.read_excel("ASHIRNR5.xlsx")
Jordan = pd.read_excel("ASHJORR5.xlsx")
Netherlands = pd.read_excel("ASHNLDR5.xlsx")
Sweden = pd.read_excel("ASHSWER5.xlsx")

HC is the home context data. This is a questionnaire filled in by the parents of each student. It contains crucial information on the immigration status of students and the languages they speak.

In [2]:
columns_to_keep = ['IDCNTRY','IDSTUD','ASBH02A','ASBH02B','ASBH03A','ASBH04', 'ASBH15A','ASBH15B','ASBH16','ASBH17A','ASBH17B','ASBH18AA','ASBH18AB']

In [4]:
df = df[columns_to_keep]

In [3]:
df = pd.concat([Turkey, Austria, Germany, Egypt, France, Iran, Jordan, Netherlands, Sweden])

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50367 entries, 0 to 5174
Data columns (total 13 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   IDCNTRY   50367 non-null  int64 
 1   IDSTUD    50367 non-null  int64 
 2   ASBH02A   41503 non-null  object
 3   ASBH02B   16269 non-null  object
 4   ASBH03A   41179 non-null  object
 5   ASBH04    41231 non-null  object
 6   ASBH15A   37405 non-null  object
 7   ASBH15B   34597 non-null  object
 8   ASBH16    40406 non-null  object
 9   ASBH17A   36503 non-null  object
 10  ASBH17B   32361 non-null  object
 11  ASBH18AA  40161 non-null  object
 12  ASBH18AB  35131 non-null  object
dtypes: int64(2), object(11)
memory usage: 5.4+ MB


It is interesting here that when the dataframes were concatenated the Dtype of the columns changed from int to object. I believe this is because the data for some countries has been inputted as strings and for other countries as int. 

ASBH02A	GEN\CHILD BORN IN <COUNTRY>	HQ-02A	Nominal	1	0			 1: Yes; 2: No	9: Omitted or invalid; Sysmis: Not administered

ASBH02B	GEN\AGE OF CHILD WHEN CAME TO <COUNTRY>	HQ-02B	Ordinal	1	0			 1: Younger than 3 years old; 2: 3 to 5 years old; 3: 6 to 7 years old; 4: 8 years old or older	9: Omitted or invalid; 6: Logically not applicable; Sysmis: Not administered

ASBH03A	GEN\LANGUAGE\<LANGUAGE OF TEST>	HQ-03a	Nominal	1	0			 1: Yes; 2: No	9: Omitted or invalid; Sysmis: Not administered

ASBH04	GEN\HOW OFTEN <LANG OF TEST> AT HOME	HQ-04	Ordinal	1	0			 1: Always; 2: Almost always; 3: Sometimes; 4: Never	9: Omitted or invalid; Sysmis: Not administered

ASBH15A	GEN\LVL OF EDUCATION\GUARDIAN A	HQ-15a	Ordinal	2	0			 1: Did not go to school; 10: Not applicable; 2: Some <Primary education—ISCED Level 1 or Lower secondary education—ISCED Level 2>; 3: <Lower secondary education—ISCED Level 2>; 4: <Upper secondary education—ISCED Level 3>; 5: <Post-secondary, non-tertiary education— ISCED Level 4>; 6: <Short-cycle tertiary education—ISCED Level 5>; 7: <Bachelor’s or equivalent level—ISCED Level 6>; 8: <Master’s or equivalent level—ISCED Level 7>; 9: <Doctor or equivalent level—ISCED Level 8>	999: Omitted or invalid; Sysmis: Not administered

ASBH15B	GEN\LVL OF EDUCATION\GUARDIAN B	HQ-15b	Ordinal	2	0			 1: Did not go to school; 10: Not applicable; 2: Some <Primary education—ISCED Level 1 or Lower secondary education—ISCED Level 2>; 3: <Lower secondary education—ISCED Level 2>; 4: <Upper secondary education—ISCED Level 3>; 5: <Post-secondary, non-tertiary education— ISCED Level 4>; 6: <Short-cycle tertiary education—ISCED Level 5>; 7: <Bachelor’s or equivalent level—ISCED Level 6>; 8: <Master’s or equivalent level—ISCED Level 7>; 9: <Doctor or equivalent level—ISCED Level 8>	999: Omitted or invalid; Sysmis: Not administered

SC is the Student Context data which is a questionnaire filled in by the students themselves as well as some totals on their achievement in the test. 

In [6]:
TurkeySC = pd.read_excel("ASGTURR5.xlsx")
AustriaSC = pd.read_excel("ASGAUTR5.xlsx")
EgyptSC = pd.read_excel("ASGEGYR5.xlsx")
FranceSC = pd.read_excel("ASGFRAR5.xlsx")
IranSC = pd.read_excel("ASGIRNR5.xlsx")
JordanSC = pd.read_excel("ASGJORR5.xlsx")
NetherlandsSC = pd.read_excel("ASGNLDR5.xlsx")
SwedenSC = pd.read_excel("ASGSWER5.xlsx")
GermanySC = pd.read_excel("ASGDEUR5.xlsx")

In [7]:
SCdf = pd.concat([TurkeySC, AustriaSC, EgyptSC, FranceSC, IranSC, JordanSC, NetherlandsSC, SwedenSC, GermanySC])

In [8]:
SCdf = SCdf[['IDCNTRY',
 'IDSTUD','ASBG01',
 'ASBG03','ASBG10C',
 'ASBG10D',
 'ASBG10E',
 'ASBG10F',
 'ASBG11A',
 'ASBG11B',
 'ASBG11C',
 'ASBG11D',
 'ASBG11E',
 'ASBG11F',
 'ASBG11G',
 'ASBG11H',
 'ASBG11I',
 'ASBG11J','ASDAGE','ASRREA01',
 'ASRREA02',
 'ASRREA03',
 'ASRREA04',
 'ASRREA05',
 'ASRLIT01',
 'ASRLIT02',
 'ASRLIT03',
 'ASRLIT04',
 'ASRLIT05',
 'ASRINF01',
 'ASRINF02',
 'ASRINF03',
 'ASRINF04',
 'ASRINF05',
 'ASRIIE01',
 'ASRIIE02',
 'ASRIIE03',
 'ASRIIE04',
 'ASRIIE05',
 'ASRRSI01',
 'ASRRSI02',
 'ASRRSI03',
 'ASRRSI04',
 'ASRRSI05']]

In [9]:
DF = pd.merge(df, SCdf, on=None)

In [10]:
def populate_column(row):
    if row['ASBH02B'] == 1:
        return 0
    elif row['ASBH02B'] == 2:
        return 3
    elif row['ASBH02B'] == 3:
        return 6
    elif row['ASBH02B'] == 4:
        return 8
    else:
        return None  # Or any default value

# Apply the function to each row
DF['MINAGEARRIVAL'] = DF.apply(populate_column, axis=1)

In [11]:
DF['MINAGEARRIVAL'].value_counts()

0.0    182
3.0    160
6.0    108
8.0     63
Name: MINAGEARRIVAL, dtype: int64

In [12]:
DF['rounded_ages'] = DF['ASDAGE'].round()

In [13]:
DF['rounded_ages'].value_counts()

10.0    38410
11.0     9104
9.0      1964
12.0      758
13.0       83
8.0        18
14.0       17
99.0        3
7.0         1
Name: rounded_ages, dtype: int64

First thing I need to do is recognise these countries and add these values to that column. 

In [14]:
DF= DF.rename(columns={'IDCNTRY': 'Country'})

# Mapping dictionary
update_dict = {40: 'Austria', 818: 'Egypt', 250:'France', 276:'Germany',364:'Iran',400:'Jordan',528:'Netherlands',752:'Sweden',792:'Turkey'}

# Updating the column using map
DF['Country'] = DF['Country'].map(update_dict).fillna(DF['Country'])

In [15]:
DF['Country'].value_counts()

Egypt          7979
Jordan         6150
Turkey         6032
Iran           5962
France         5339
Sweden         5175
Austria        4806
Germany        4611
Netherlands    4313
Name: Country, dtype: int64

In [17]:
DF['ASBH02A'].value_counts()

Yes    25072
1.0    13403
No      2342
2.0      548
9.0      138
Name: ASBH02A, dtype: int64

In [18]:
# Mapping dictionary
update_dict2 = {1.0: 'Yes',2.0:'No' }

# Updating the column using map
DF['ASBH02A'] = DF['ASBH02A'].map(update_dict2).fillna(DF['ASBH02A'])
DF['ASBH02A'].value_counts()

Yes    38475
No      2890
9.0      138
Name: ASBH02A, dtype: int64

So we have 2890 children who were not born in the country where they took the test.

In [19]:
DF["ASBH03A"].value_counts()

Yes    25915
1.0    12270
2.0     1477
No      1175
9.0      342
Name: ASBH03A, dtype: int64

In [21]:
# Updating the column using map
DF['ASBH03A'] = DF['ASBH03A'].map(update_dict2).fillna(DF['ASBH03A'])
DF['ASBH03A'].value_counts()

Yes    38185
No      2652
9.0      342
Name: ASBH03A, dtype: int64

And we have 2652 children who did not speak the language of the test before they started school. So some children were born elsewhere but were already speaking the language of the test before they started school.

I want to see how many columns are affected by having strings and ints - so I want to see th unique values for each column. 

In [30]:
# Apply unique() to each column
unique_values = DF.apply(lambda col: col.unique())

print(unique_values)

Country                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 

In [34]:
ordinal_mapping_frequency = {
    'At least once a week': 1,
    'Once or twice a month': 2,
    'A few times a year': 3,
    'Never': 4
}

In [32]:
experience_in_school_columns = ['ASBG11A',
 'ASBG11B',
 'ASBG11C',
 'ASBG11D',
 'ASBG11E',
 'ASBG11F',
 'ASBG11G',
 'ASBG11H',
 'ASBG11I',
 'ASBG11J']

In [35]:
for column in experience_in_school_columns:
    DF[column]= DF[column].map(ordinal_mapping_frequency)

In [29]:
pd.set_option('display.max_columns', None)  # Display all columns
pd.set_option('display.max_colwidth', None)  # Optional: Set max column width
pd.set_option('display.expand_frame_repr', False)  # Optional: Do not wrap line

In [43]:
# Apply unique() to each column
unique_values_school_experience = DF[experience_in_school_columns].apply(lambda col: col.unique())

print(unique_values_school_experience)

   ASBG11A  ASBG11B  ASBG11C  ASBG11D  ASBG11E  ASBG11F  ASBG11G  ASBG11H  ASBG11I  ASBG11J
0      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN
1      4.0      4.0      3.0      2.0      4.0      4.0      4.0      3.0      4.0      4.0
2      2.0      2.0      1.0      4.0      3.0      3.0      2.0      4.0      1.0      3.0
3      3.0      3.0      2.0      3.0      2.0      2.0      3.0      1.0      3.0      1.0
4      1.0      1.0      4.0      1.0      1.0      1.0      1.0      2.0      2.0      2.0


In [40]:
ordinal_mapping_agreement = { 
    'Agree a lot': 1,
    'Agree a little': 2,
    'Disagree a little': 3,
    'Disagree a lot': 4
}

In [39]:
school_feeling_columns = ['ASBG10C',
 'ASBG10D',
 'ASBG10E',
 'ASBG10F']

In [41]:
for column in school_feeling_columns:
    DF[column]= DF[column].map(ordinal_mapping_agreement)

In [44]:
# Apply unique() to each column
unique_values_school_feeling = DF[school_feeling_columns].apply(lambda col: col.unique())

print(unique_values_school_feeling)

   ASBG10C  ASBG10D  ASBG10E  ASBG10F
0      NaN      NaN      NaN      NaN
1      2.0      2.0      3.0      1.0
2      1.0      1.0      1.0      2.0
3      3.0      3.0      2.0      3.0
4      4.0      4.0      4.0      4.0


In [ ]:
ordinal_mapping_frequency = {
    'At least once a week': 1,
    'Once or twice a month': 2,
    'A few times a year': 3,
    'Never': 4
}

In [45]:
ordinal_mapping_education_levels = {
1: 'Did not go to school', 
 2: 'Some <Primary education—ISCED Level 1 or Lower secondary education—ISCED Level 2>', 
 3: '<Lower secondary education—ISCED Level 2>', 
 4: '<Upper secondary education—ISCED Level 3>', 
 5: '<Post-secondary, non-tertiary education— ISCED Level 4>', 
 6: '<Short-cycle tertiary education—ISCED Level 5>', 
 7: '<Bachelor’s or equivalent level—ISCED Level 6>', 
 8: '<Master’s or equivalent level—ISCED Level 7>', 
 9: '<Doctor or equivalent level—ISCED Level 8>',
 10: 'Not applicable'
 }

In [46]:
# Reverse the mapping dictionary
reverse_mapping_education_levels = {v: k for k, v in ordinal_mapping_education_levels.items()}

In [47]:
DF['ASBH15A'] = DF['ASBH15A'].map(reverse_mapping_education_levels)

In [49]:
DF['ASBH15B'] = DF['ASBH15B'].map(reverse_mapping_education_levels)

In [51]:
 ordinal_mapping_profession_levels = {
 1: 'Has never worked for pay',
 10: 'Professional Includes scientists,mathematicians; computer scientists,architects; engineers; life science and health pr',
 11: 'Technician or Associate Professional Includes science, engineering, and computer associates and technicians; life scienc',
 12: 'Not applicable',
 2: 'Small Business Owner Includes owners of small businesses (fewer than 25 employees) such as retail shops, services, resta',
 3: 'Clerical Worker Includes office clerks; secretaries; data entry operators; customer service clerks',
 4: 'Service or Sales Worker Includes travel attendants; restaurant service workers; personal care workers; protective servic',
 5: 'Skilled Agricultural or Fishery Worker Includes farmers; forestry workers; fishery workers',
 6: 'Craft or Trade Worker Includes builders, carpenters, plumbers, electricians, metal workers; machine mechanics; handicraf',
 7: 'Plant or Machine Operator Includes plant and machine operators; assembly-line operators; motor-vehicle drivers',
 8: 'General'
 }

In [52]:
# Reverse the mapping dictionary
reverse_mapping_profession_levels = {v: k for k, v in ordinal_mapping_profession_levels.items()}

In [53]:
DF['ASBH17A'] = DF['ASBH17A'].map(reverse_mapping_profession_levels)

In [55]:
DF['ASBH17B'] = DF['ASBH17B'].map(reverse_mapping_profession_levels)

Remember to git once you've done all these value changes.

In [58]:
ordinal_mapping_language = {
     1: 'I always speak <language of test> at home',
    2: 'I almost always speak <language of test> at home',
    3: 'I sometimes speak <language of test> and sometimes speak another language at home',
    4: 'I never speak <language of test> at home'
    
}

In [65]:
# Reverse the mapping dictionary
reverse_mapping_language = {v: k for k, v in ordinal_mapping_language.items()}
DF['ASBG03'] = DF['ASBG03'].map(reverse_mapping_language)
DF['ASBG03'].unique()

array([nan,  1.,  3.,  2.,  4.])

In [62]:
ordinal_mapping_language_parents = {
    1: 'Always',
    2: 'Almost always',
    3: 'Sometimes',
    4: 'Never'
}

In [63]:
# Reverse the mapping dictionary
reverse_mapping_language_parents = {v: k for k, v in ordinal_mapping_language_parents.items()}
DF['ASBH04'] = DF['ASBH04'].map(reverse_mapping_language_parents)
DF['ASBH04'].unique()

array([nan,  1.,  4.,  2.,  3.])

In [66]:
# Apply unique() to each column
unique_values = DF.apply(lambda col: col.unique())

print(unique_values)

Country                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 

In [ ]:
ASBH18AA                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    [9.0, 1.0, 2.0, nan, 6.0, Yes, No]
ASBH18AB                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    [1.0, 9.0, 2.0, nan, 6.0, Yes, No]
ASBG01                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     [2.0, 1.0, nan, Girl, Boy, <Other>]

In [68]:
DF.to_excel('Not_fully_clean.xlsx')

In [70]:
DF2= pd.read_excel('Not_fully_clean.xlsx')

In [83]:
# Define the function to map patterns to integers
def map_using_patterns(text):
    # Dictionary of regex patterns to integers
    pattern_to_integer = {
        r'Lower secondary education': 1,
        r'Upper secondary education': 2,
        r'Post-secondary, non-tertiary education': 3,
        r'Short-cycle tertiary education': 4,
        r'Bachelor’s or equivalent level': 5,
        r'Postgraduate degree': 6
    }

    # Check if the input is a string
    if isinstance(text, str):
        for pattern, integer in pattern_to_integer.items():
            if pd.Series([text]).str.contains(pattern).any():
                return integer
    return text  # Return the original value if it's not a string or no pattern matches

# Apply the function to the 'ASBH16' column
DF2['ASBH16'] = DF2['ASBH16'].apply(map_using_patterns)


In [85]:
DF2['ASBH16'].value_counts()

6.0    16285
5.0    11132
4.0     4655
2.0     4523
3.0     2452
1.0     1093
9.0      266
Name: ASBH16, dtype: int64

In [87]:
mapping_dict = {"Yes": 1, "No": 2}
DF2['ASBH18AA']= DF2['ASBH18AA'].map(mapping_dict)
DF2['ASBH18AB']= DF2['ASBH18AB'].map(mapping_dict)
DF2['ASBH18AA'].unique()
DF2['ASBH18AB'].unique()

array([nan,  1.,  2.])

In [88]:
mapping_dict_gender = {"Girl": 1, "Boy": 2, "<Other>": 3}
DF2['ASBG01']= DF2['ASBG01'].map(mapping_dict_gender)
DF2['ASBG01'].unique()

array([nan,  1.,  2.,  3.])

In [89]:
# Apply unique() to each column
unique_values_DF2 = DF2.apply(lambda col: col.unique())

print(unique_values_DF2)

Unnamed: 0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71

do I need to change the ones ASBH02A etc.? I'm not sure. Shall we do some practice typing? Because actually I don't do that much typing and I think I'm losing it a little bit. Which isn't good. So I am going to go home now and go to tesco on the way. I'm going to look for lettuce and cottage cheese and compare the carb content of cottage cheese to that of cream cheese etc. Because I could potentially use cottage cheese as a dip. Perhaps with the addition of of avocado. I need to not pre-empt weekend snacking. Perhaps if I go to Norwich on Saturdays that might help my feeling and then see Jake or Rosie in the afternoon. Because I can't do Saturdays at home! Have to get out of the house. And then Sunday could be my day of rest. Good typing gal! It's time to go, isn't it? Can do some reading in the garden in the afternoon - that could be nice.! Cool, let's go.